File: src/pkg/player/8durt/lib/Span.lua
-- package: lib

require("ktox-lib")
ktox_sourcemap_traceback(debug and debug.getinfo and (debug.getinfo(1) or {}).short_src or "", "lib/Span.kt", {["1-41"]=1,["42-49"]=10,["50"]=22,["51-52"]=23,["53-55"]=25}, "lib")

---@class IntSpan
---@field start number
---@field finish number
IntSpan = {}
IntSpan.__index = IntSpan

function IntSpan:new(start, finish)
    local self = setmetatable({}, IntSpan)
    self.start = start
    self.finish = finish
    return self
end

function IntSpan:equals(other)
    return self.start == other.start and self.finish == other.finish
end
IntSpan.__eq = function(a, b) return a:equals(b) end
function IntSpan:toString()
    return "IntSpan(" .. "start=" .. tostring(self.start) .. ", " .. "finish=" .. tostring(self.finish) .. ")"
end
IntSpan.__tostring = function(a) return a:toString() end
function IntSpan:copy(start, finish)
    if start == nil then start = self.start end
    if finish == nil then finish = self.finish end
    return IntSpan:new(start, finish)
end
function IntSpan:component1()
    return self.start
end
function IntSpan:component2()
    return self.finish
end

---@param span IntSpan
---@return number
function stepFor(span)
    return (span.start <= span.finish and 1 or -1)
end

---@param current number
---@param limit number
---@param step number
---@return boolean
function pastEnd(current, limit, step)
    if step > 0 then
        return current > limit
    end
    return current < limit
end

File: ktox-lua-cc/src/main/kotlin/lib/Span.kt
package lib

// An inclusive [start, finish] integer range that preserves direction —
// unlike a sorted min:max pair, start > finish is a valid descending span.
// Shared by any program that sweeps a bounded x/y/z footprint (Digsite,
// ExcavatePro).
data class IntSpan(val start: Int, val finish: Int)

fun stepFor(span: IntSpan): Int {
    return if (span.start <= span.finish) 1 else -1
}

// True once `current` has moved past `limit` while stepping by `step`
// (whose sign gives the direction of travel).
//
// Written as an if/else block, NOT `return if (cond) A else B` — ktox
// compiles that single-expression form to Lua's `(cond and A or B)`
// idiom, which is broken here: A (`current > limit`) is itself a
// boolean, and whenever it's false, `cond and false` is false, so Lua
// falls through to B regardless of cond. See AGENTS.md.
fun pastEnd(current: Int, limit: Int, step: Int): Boolean {
    if (step > 0) {
        return current > limit
    }
    return current < limit
}
